From 68c2ffa65eeae939beaa6cf7f803155fecad25c8 Mon Sep 17 00:00:00 2001
From: Marc Brakels <marc.brakels@topic.nl>
Date: Fri, 5 Mar 2021 15:47:49 +0100
Subject: [PATCH 1/2] UBI support u-boot

---
 Kconfig                       |  3 +++
 include/configs/zynq-common.h | 26 ++++++++++++++++++++++++++
 2 files changed, 29 insertions(+)

diff --git a/Kconfig b/Kconfig
index d80be6e218..d5d2ed8c2f 100644
--- a/Kconfig
+++ b/Kconfig
@@ -571,6 +571,9 @@ config SYS_TEXT_BASE
 	help
 	  The address in memory that U-Boot will be running from, initially.
 
+config KERNEL_ARGS
+	string "Extra bootargs options"
+	default " " 
 
 
 config SYS_CLK_FREQ
diff --git a/include/configs/zynq-common.h b/include/configs/zynq-common.h
index d0614d8da4..06af9b307d 100644
--- a/include/configs/zynq-common.h
+++ b/include/configs/zynq-common.h
@@ -283,3 +283,29 @@
 #define CONFIG_SYS_UBOOT_START	CONFIG_SYS_TEXT_BASE
 
 #endif /* __CONFIG_ZYNQ_COMMON_H */
+
+#ifndef CONFIG_KERNEL_ARGS
+#define CONFIG_KERNEL_ARGS ""
+#endif
+
+
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#define CONFIG_EXTRA_ENV_SETTINGS	\
+	"scriptaddr=0x20000\0"	\
+	"script_size_f=0x40000\0"	\
+	"fdt_addr_r=0x1f00000\0"        \
+	"pxefile_addr_r=0x2000000\0"    \
+	"kernel_addr_r=0x2000000\0"     \
+	"scriptaddr=0x3000000\0"        \
+	"ramdisk_addr_r=0x3100000\0"    \
+	DFU_ALT_INFO \
+	BOOTENV \
+	"mtdids=" CONFIG_MTDIDS_DEFAULT "\0" \
+	"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" \
+	"kernelargs=" CONFIG_KERNEL_ARGS "\0" \
+	"bootcmd_ubi=run ubi_mount && ubifsload ${scriptaddr} /boot/boot.scr && source ${scriptaddr}\0"    \
+  "scriptaddr=0x1900000\0"   \
+	"ubi_mount=sf probe && ubi part qspi-rootfs && ubifsmount ubi0:qspi-rootfs && setenv devtype ubi\0" \
+	"bootubipart=qspi-rootfs\0" \
+	"bootubivol=qspi-rootfs\0" \
+	"bootcmd_qspi=run bootcmd_ubi\0"
-- 
2.17.1

