From 1c85787c1aace97906e37a56212966625245710a Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Thu, 27 Jul 2023 12:16:42 +0200
Subject: [PATCH] zynqmp: Boot with UBI from QSPI

Upstream-Status: Inappropriate
---
 include/configs/xilinx_zynqmp.h | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/include/configs/xilinx_zynqmp.h b/include/configs/xilinx_zynqmp.h
index 5bc117e190..2910ef3fad 100644
--- a/include/configs/xilinx_zynqmp.h
+++ b/include/configs/xilinx_zynqmp.h
@@ -108,10 +108,7 @@
 #endif
 
 #define BOOTENV_DEV_QSPI(devtypeu, devtypel, instance) \
-	"bootcmd_" #devtypel #instance "=sf probe " #instance " 0 0 && " \
-		       "sf read $scriptaddr $script_offset_f $script_size_f && " \
-		       "echo QSPI: Trying to boot script at ${scriptaddr} && " \
-		       "source ${scriptaddr}; echo QSPI: SCRIPT FAILED: continuing...;\0"
+	"bootcmd_" #devtypel #instance "=bootubipart=qspi-rootfs; bootubioff=0; bootubivol=qspi-rootfs; run ubifs_boot;\0"
 
 #define BOOTENV_DEV_NAME_QSPI(devtypeu, devtypel, instance) \
 	#devtypel #instance " "
-- 
2.43.0

