From 4592f636f644ec695be1e32245a01b10624cb727 Mon Sep 17 00:00:00 2001
From: Leon Leijssen <leon.leijssen@topic.nl>
Date: Thu, 9 Jan 2020 11:20:08 +0100
Subject: [PATCH 12/13] Add usb reset for miamimp

---
 include/configs/topic_miamimp.h | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/include/configs/topic_miamimp.h b/include/configs/topic_miamimp.h
index 7888490fe1..ff9b892ee7 100644
--- a/include/configs/topic_miamimp.h
+++ b/include/configs/topic_miamimp.h
@@ -57,9 +57,13 @@
 	"sdroot0=setenv bootargs $bootargs root=/dev/mmcblk0p2 rw rootwait\0" \
 	"sdroot1=setenv bootargs $bootargs root=/dev/mmcblk1p2 rw rootwait\0" \
 	"usb_dfu_spl=booti $kernel_addr - $fdt_addr\0" \
-	"usbhostboot=usb start && load usb 0 $fdt_addr system.dtb && " \
+	"usb_hostboot=usb start && load usb 0 $fdt_addr system.dtb && " \
 		"load usb 0 $kernel_addr Image && " \
-		"booti $kernel_addr - $fdt_addr\0"
+		"booti $kernel_addr - $fdt_addr\0" \
+	"usb_reset=gpio clear gpio@41_0 && gpio set gpio@41_0\0" \
+	"usb_reset_boot=run usb_reset; run usb_boot\0" \
+	"usb_reset_hostboot=run usb_reset; run usb_hostboot\0" \
+	"usb_reset_dfu_ram=run usb_reset; run dfu_ram\0"
 
 #include "xilinx_zynqmp.h"
 
-- 
2.17.1

