From 2c9d1ea87e42f9ef06cb80c0237bb6ebe20f8d57 Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Tue, 25 Feb 2014 15:39:55 +0100
Subject: [PATCH] Configure clock as 32MHz on zynq-miami

---
 include/configs/zynq_miami.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/zynq_miami.h b/include/configs/zynq_miami.h
index d4113aa..6e2826e 100644
--- a/include/configs/zynq_miami.h
+++ b/include/configs/zynq_miami.h
@@ -11,7 +11,7 @@
 #define __CONFIG_ZYNQ_MIAMI_H
 
 #define CONFIG_SYS_SDRAM_SIZE		(512 * 1024 * 1024)
-
+#define CONFIG_ZYNQ_PS_CLK_FREQ		32000000UL
 #define CONFIG_ZYNQ_SERIAL_UART1
 /* Ethernet needs external clock, not really supported here yet but
  * it won't compile without it. */
-- 
1.7.9.5

