From 88776249ca056b8b4ca7550713bbc28ac077b327 Mon Sep 17 00:00:00 2001
From: Philip Balister <philip@opensdr.com>
Date: Fri, 5 Oct 2012 10:54:10 -0400
Subject: [PATCH] Change boot options so linux can boot without a ramdisk.

Signed-off-by: Philip Balister <philip@opensdr.com>
---
 include/configs/zynq_common.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/zynq_common.h b/include/configs/zynq_common.h
index 5c6086c..6f5ed17 100644
--- a/include/configs/zynq_common.h
+++ b/include/configs/zynq_common.h
@@ -61,7 +61,7 @@
 		"fatload mmc 0 0x3000000 ${kernel_image};" \
 		"fatload mmc 0 0x2A00000 ${devicetree_image};" \
 		"fatload mmc 0 0x2000000 ${ramdisk_image};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0" \
+		"bootm 0x3000000 - 0x2A00000\0" \
 	"nandboot=echo Copying Linux from NAND flash to RAM...;" \
 		"nand read 0x3000000 0x200000 ${nand_kernel_size};" \
 		"nand read 0x2A00000 0x700000 0x20000;" \
-- 
1.7.11.4

