From 85d8876bcef07e737302d6f4b79f791e625bb14e Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Wed, 19 Feb 2014 15:18:47 +0100
Subject: [PATCH 02/10] Disable CONFIG_SPL_OS_BOOT

---
 include/configs/zynq-common.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/configs/zynq-common.h b/include/configs/zynq-common.h
index d601426..9fe5c82 100644
--- a/include/configs/zynq-common.h
+++ b/include/configs/zynq-common.h
@@ -504,7 +504,7 @@
 #define CONFIG_SYS_SPI_ARGS_OFFS	0 /* FIXME */
 #define CONFIG_SYS_SPI_ARGS_SIZE	0 /* FIXME */
 
-#define CONFIG_SPL_FAT_LOAD_ARGS_NAME		"system.dtb"
+#define CONFIG_SPL_FAT_LOAD_ARGS_NAME		"devicetree.dtb"
 #define CONFIG_SPL_FAT_LOAD_KERNEL_NAME		"uImage"
 
 /* Not using MMC raw mode - just for compilation purpose */
@@ -531,7 +531,7 @@
 #endif
 
 /* for booting directly linux */
-#define CONFIG_SPL_OS_BOOT
+#undef CONFIG_SPL_OS_BOOT
 #define CONFIG_SYS_SPI_KERNEL_OFFS	0 /* FIXME */
 
 /* SP location before relocation, must use scratch RAM */
-- 
1.9.1

