From ca9faba6323bf31e0802d7d2aee8aaba23a01c1c Mon Sep 17 00:00:00 2001
From: Mike Looijmans <mike.looijmans@topic.nl>
Date: Fri, 21 Mar 2014 08:15:48 +0100
Subject: [PATCH 05/10] Load FPGA from "big" u-boot instead of SPL

---
 include/configs/zynq-common.h |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/include/configs/zynq-common.h b/include/configs/zynq-common.h
index 3f1a7e9..8731721 100644
--- a/include/configs/zynq-common.h
+++ b/include/configs/zynq-common.h
@@ -278,7 +278,9 @@
 #  define CONFIG_ENV_IS_NOWHERE
 # endif
 
+# ifndef CONFIG_ENV_SECT_SIZE
 # define CONFIG_ENV_SECT_SIZE		CONFIG_ENV_SIZE
+# endif
 # ifndef CONFIG_ENV_OFFSET
 #  define CONFIG_ENV_OFFSET		0x60000
 # endif
@@ -304,7 +306,7 @@
 	"ramdisk_load_address=0x4000000\0"	\
 	"devicetree_image=devicetree.dtb\0"	\
 	"devicetree_load_address=0x2000000\0"	\
-	"bitstream_image=system.bit.bin\0"	\
+	"bitstream_image=fpga.bin\0"	\
 	"bootscript=" CONFIG_EXTRA_ENV_BOOTSCRIPT "\0" \
 	"loadbit_addr=0x100000\0"	\
 	"loadbootenv_addr=0x2000000\0" \
@@ -477,7 +479,6 @@
 #define CONFIG_SPL_LDSCRIPT	"arch/arm/cpu/armv7/zynq/u-boot-spl.lds"
 
 /* FPGA support */
-#define CONFIG_SPL_FPGA_SUPPORT
 #define CONFIG_SPL_FPGA_LOAD_ADDR      0x1000000
 /* #define CONFIG_SPL_FPGA_BIT */
 #ifdef CONFIG_SPL_FPGA_BIT
-- 
1.7.9.5

