Index: git/include/configs/zynq_common.h
===================================================================
--- git.orig/include/configs/zynq_common.h
+++ git/include/configs/zynq_common.h
@@ -32,55 +32,47 @@
  */
 #define CONFIG_OF_LIBFDT
 
+/* Enables if..then..else constructs */
+#define CONFIG_SYS_HUSH_PARSER
+#define CONFIG_SYS_PROMPT_HUSH_PS2 "> "
+
 /* Default environment */
 #define CONFIG_EXTRA_ENV_SETTINGS	\
 	"ethaddr=00:0a:35:00:01:22\0"	\
 	"kernel_image=uImage\0"	\
-	"ramdisk_image=uramdisk.image.gz\0"	\
 	"devicetree_image=devicetree.dtb\0"	\
+	"ramdisk_image=uramdisk.image.gz\0"	\
 	"kernel_addr=0x79C0\0"	\
 	"kernel_size=0x140000\0"	\
-	"ramdisk_size=0x200000\0"	\
 	"nand_kernel_size=0x400000\0"	\
-	"nand_ramdisk_size=0x400000\0"	\
 	"fdt_high=0x20000000\0"	\
 	"initrd_high=0x20000000\0"	\
 	"norboot=echo Copying Linux from NOR flash to RAM...;" \
-		"cp 0xE2100000 ${kernel_addr} ${kernel_size};" \
-		"cp 0xE2600000 0x2A00000 0x20000;" \
-		"echo Copying ramdisk...;" \
-		"cp 0xE3000000 0x2000000 ${ramdisk_size};" \
-		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
+		"bootm 0xE2100000 - 0xE2600000\0" \
 	"qspiboot=echo Copying Linux from QSPI flash to RAM...;" \
-		"cp 0xFC100000 ${kernel_addr} ${kernel_size};" \
-		"cp 0xFC600000 0x2A00000 0x20000;" \
-		"echo Copying ramdisk...;" \
-		"cp 0xFC800000 0x2000000 ${ramdisk_size};" \
-		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
+		"bootm 0xFC080000 - 0xFC070000\0" \
 	"sdboot=echo Copying Linux from SD to RAM...;" \
 		"mmcinfo;" \
 		"fatload mmc 0 ${kernel_addr} ${kernel_image};" \
 		"fatload mmc 0 0x2A00000 ${devicetree_image};" \
-		"fatload mmc 0 0x2000000 ${ramdisk_image};" \
 		"bootm ${kernel_addr} - 0x2A00000\0" \
 	"nandboot=echo Copying Linux from NAND flash to RAM...;" \
 		"nand read ${kernel_addr} 0x200000 ${nand_kernel_size};" \
 		"nand read 0x2A00000 0x700000 0x20000;" \
-		"echo Copying ramdisk...;" \
-		"nand read 0x2000000 0x900000 ${nand_ramdisk_size};" \
-		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
+		"bootm ${kernel_addr} - 0x2A00000\0" \
 	"jtagboot=echo TFTPing Linux to RAM...;" \
 		"tftp ${kernel_addr} ${kernel_image};" \
 		"tftp 0x2A00000 ${devicetree_image};" \
 		"tftp 0x2000000 ${ramdisk_image};" \
 		"bootm ${kernel_addr} 0x2000000 0x2A00000\0"
 
-/* default boot is according to the bootmode switch settings */
-#define CONFIG_BOOTCOMMAND "run modeboot"
+/* default boot is according to the bootmode switch settings, unless you insert a SD card with an autorun script */
+#define CONFIG_BOOTCOMMAND "if mmcinfo; then if fatload mmc 0 0x1900000 zynq-zc702-autorun.scr; then source 0x1900000; fi; fi; run modeboot"
 
 #define CONFIG_BAUDRATE		115200
 #define CONFIG_SYS_BAUDRATE_TABLE { 9600, 38400, 115200 }
-#define CONFIG_BOOTDELAY	3 /* -1 to Disable autoboot */
+#define CONFIG_ZERO_BOOTDELAY_CHECK
+#define CONFIG_BOOTDELAY	0 /* -1 to Disable autoboot */
 
 #define	CONFIG_PSS_SERIAL
 #define	CONFIG_RTC_XPSSRTC
