Index: git/include/configs/zynq_zed.h
===================================================================
--- git.orig/include/configs/zynq_zed.h
+++ git/include/configs/zynq_zed.h
@@ -36,26 +36,26 @@
 #define CONFIG_ENV_IS_NOWHERE
 
 #undef CONFIG_EXTRA_ENV_SETTINGS
-#define CONFIG_EXTRA_ENV_SETTINGS 	\
+#define CONFIG_EXTRA_ENV_SETTINGS	\
+	"bootscript=zedboard-autorun.scr\0"	\
 	"ethaddr=00:0a:35:00:01:22\0"	\
-	"kernel_size=0x140000\0" 	\
-	"ramdisk_size=0x200000\0" 	\
-	"qspiboot=sf probe 0 0 0;" \
-		"sf read 0x8000 0x100000 0x2c0000;" \
-		"sf read 0x1000000 0x3c0000 0x40000;" \
-		"sf read 0x800000 0x400000 0x800000;" \
-		"go 0x8000\0" \
+	"kernel_image=uImage\0"	\
+	"devicetree_image=devicetree.dtb\0"	\
+	"ramdisk_image=uramdisk.image.gz\0"	\
+	"kernel_addr=0x79C0\0"	\
+	"kernel_size=0x400000\0"	\
+	"qspiboot=echo Booting Linux from QSPI flash...;" \
+		"bootm 0xFC080000 - 0xFC070000\0" \
 	"sdboot=echo Copying Linux from SD to RAM...;" \
 		"mmcinfo;" \
-		"fatload mmc 0 0x8000 zImage;" \
-		"fatload mmc 0 0x1000000 devicetree.dtb;" \
-		"fatload mmc 0 0x800000 ramdisk8M.image.gz;" \
-		"go 0x8000\0" \
+		"fatload mmc 0 ${kernel_addr} ${kernel_image};" \
+		"fatload mmc 0 0x2A00000 ${devicetree_image};" \
+		"bootm ${kernel_addr} - 0x2A00000\0" \
 	"jtagboot=echo TFTPing Linux to RAM...;" \
-		"tftp 0x8000 zImage;" \
-		"tftp 0x1000000 devicetree.dtb;" \
-		"tftp 0x800000 ramdisk8M.image.gz;" \
-		"go 0x8000\0"
+		"tftp ${kernel_addr} ${kernel_image};" \
+		"tftp 0x2A00000 ${devicetree_image};" \
+		"tftp 0x2000000 ${ramdisk_image};" \
+		"bootm ${kernel_addr} 0x2000000 0x2A00000\0"
 
 #include <config_cmd_default.h>
 #define CONFIG_CMD_DATE		/* RTC? */
