Index: git/include/configs/zynq_common.h
===================================================================
--- git.orig/include/configs/zynq_common.h
+++ git/include/configs/zynq_common.h
@@ -38,6 +38,7 @@
 	"kernel_image=uImage\0"	\
 	"ramdisk_image=uramdisk.image.gz\0"	\
 	"devicetree_image=devicetree.dtb\0"	\
+	"kernel_addr=0x79C0\0"	\
 	"kernel_size=0x140000\0"	\
 	"ramdisk_size=0x200000\0"	\
 	"nand_kernel_size=0x400000\0"	\
@@ -45,34 +46,34 @@
 	"fdt_high=0x20000000\0"	\
 	"initrd_high=0x20000000\0"	\
 	"norboot=echo Copying Linux from NOR flash to RAM...;" \
-		"cp 0xE2100000 0x3000000 ${kernel_size};" \
+		"cp 0xE2100000 ${kernel_addr} ${kernel_size};" \
 		"cp 0xE2600000 0x2A00000 0x20000;" \
 		"echo Copying ramdisk...;" \
 		"cp 0xE3000000 0x2000000 ${ramdisk_size};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0" \
+		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
 	"qspiboot=echo Copying Linux from QSPI flash to RAM...;" \
-		"cp 0xFC100000 0x3000000 ${kernel_size};" \
+		"cp 0xFC100000 ${kernel_addr} ${kernel_size};" \
 		"cp 0xFC600000 0x2A00000 0x20000;" \
 		"echo Copying ramdisk...;" \
 		"cp 0xFC800000 0x2000000 ${ramdisk_size};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0" \
+		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
 	"sdboot=echo Copying Linux from SD to RAM...;" \
 		"mmcinfo;" \
-		"fatload mmc 0 0x3000000 ${kernel_image};" \
+		"fatload mmc 0 ${kernel_addr} ${kernel_image};" \
 		"fatload mmc 0 0x2A00000 ${devicetree_image};" \
 		"fatload mmc 0 0x2000000 ${ramdisk_image};" \
-		"bootm 0x3000000 - 0x2A00000\0" \
+		"bootm ${kernel_addr} - 0x2A00000\0" \
 	"nandboot=echo Copying Linux from NAND flash to RAM...;" \
-		"nand read 0x3000000 0x200000 ${nand_kernel_size};" \
+		"nand read ${kernel_addr} 0x200000 ${nand_kernel_size};" \
 		"nand read 0x2A00000 0x700000 0x20000;" \
 		"echo Copying ramdisk...;" \
 		"nand read 0x2000000 0x900000 ${nand_ramdisk_size};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0" \
+		"bootm ${kernel_addr} 0x2000000 0x2A00000\0" \
 	"jtagboot=echo TFTPing Linux to RAM...;" \
-		"tftp 0x3000000 ${kernel_image};" \
+		"tftp ${kernel_addr} ${kernel_image};" \
 		"tftp 0x2A00000 ${devicetree_image};" \
 		"tftp 0x2000000 ${ramdisk_image};" \
-		"bootm 0x3000000 0x2000000 0x2A00000\0"
+		"bootm ${kernel_addr} 0x2000000 0x2A00000\0"
 
 /* default boot is according to the bootmode switch settings */
 #define CONFIG_BOOTCOMMAND "run modeboot"
