diff --git a/plat/xilinx/zynqmp/bl31_zynqmp_setup.c b/plat/xilinx/zynqmp/bl31_zynqmp_setup.c
index bf19147..585ad56 100644
--- a/plat/xilinx/zynqmp/bl31_zynqmp_setup.c
+++ b/plat/xilinx/zynqmp/bl31_zynqmp_setup.c
@@ -119,17 +119,12 @@ void bl31_early_platform_setup(bl31_params_t *from_bl2,
 	SET_PARAM_HEAD(&bl33_image_ep_info, PARAM_EP, VERSION_1, 0);
 	SET_SECURITY_STATE(bl33_image_ep_info.h.attr, NON_SECURE);
 
-	if (zynqmp_get_bootmode() == ZYNQMP_BOOTMODE_JTAG) {
 		/* use build time defaults in JTAG boot mode */
 		bl32_image_ep_info.pc = BL32_BASE;
 		bl32_image_ep_info.spsr = arm_get_spsr_for_bl32_entry();
 		bl33_image_ep_info.pc = plat_get_ns_image_entrypoint();
 		bl33_image_ep_info.spsr = SPSR_64(MODE_EL2, MODE_SP_ELX,
 						  DISABLE_ALL_EXCEPTIONS);
-	} else {
-		/* use parameters from FSBL */
-		fsbl_atf_handover(&bl32_image_ep_info, &bl33_image_ep_info);
-	}
 
 	NOTICE("BL31: Secure code at 0x%lx\n", bl32_image_ep_info.pc);
 	NOTICE("BL31: Non secure code at 0x%lx\n", bl33_image_ep_info.pc);
