Index: linux/drivers/gpu/drm/Kconfig
===================================================================
--- linux.orig/drivers/gpu/drm/Kconfig
+++ linux/drivers/gpu/drm/Kconfig
@@ -179,10 +179,6 @@ config DRM_ENCODER_ADV7511
 	tristate "AV7511 encoder"
 	depends on I2C && DRM
 	select REGMAP_I2C
-	select DRM_KMS_HELPER
-	select FB_SYS_FILLRECT
-	select FB_SYS_COPYAREA
-	select FB_SYS_IMAGEBLIT
 
 source "drivers/gpu/drm/analog/Kconfig"
 
