--- a/psu_init.c	2021-05-25 15:36:08.464538375 +0200
+++ b/psu_init.c	2021-05-25 16:43:10.800532630 +0200
@@ -23448,6 +23448,38 @@
 	return 0;
 }
 
+static void psu_init_sdio_pinmux(void)
+{
+	/* Check currently active boot mode, this is "5" for SD card boot */
+	if ((Xil_In32(0xFF5E0200) & 0x7) != 0x5) {
+		/* SD not detected, pinmux the WiFi chip card */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_46_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_47_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_48_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_49_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_50_OFFSET, 0x000000FEU, 0x00000000U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_51_OFFSET, 0x000000FEU, 0x00000000U);
+		/* Set feedback clock to EMIO */
+		PSU_Mask_Write(IOU_SLCR_SDIO_CLK_CTRL_OFFSET,0x00040000U, 0x00040000U);
+		/* Voltage support, fixed slot */
+		PSU_Mask_Write(IOU_SLCR_SD_CONFIG_REG2_OFFSET, 0x33840000U, 0x12800000U);
+	} else {
+		/* Set pin 45 as SD card detect, pin 46..51 to SDIO1 */
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_45_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_46_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_47_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_48_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_49_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_50_OFFSET, 0x000000FEU, 0x00000010U);
+		PSU_Mask_Write(IOU_SLCR_MIO_PIN_51_OFFSET, 0x000000FEU, 0x00000010U);
+		/* Set feedback clock to MIO 51 */
+		PSU_Mask_Write(IOU_SLCR_SDIO_CLK_CTRL_OFFSET, 0x00070000U, 0x00000000U);
+		/* Voltage support 3v3 only, removable slot */
+		PSU_Mask_Write(IOU_SLCR_SD_CONFIG_REG2_OFFSET, 0x33840000U, 0x00800000U);
+	}
+}
+
 int
 psu_init(void)
 {
@@ -23467,6 +23499,8 @@
 	status &=    psu_afi_config();
 	psu_ddr_qos_init_data();
 
+	psu_init_sdio_pinmux();
+
 	if (status == 0)
 		return 1;
 	return 0;
@@ -23489,6 +23523,8 @@
 	status &=    psu_afi_config();
 	psu_ddr_qos_init_data();
 
+	psu_init_sdio_pinmux();
+
 	if (status == 0)
 		return 1;
 	return 0;
